packages/Php70/src/Rector/FuncCall/NonVariableToVariableOnFunctionCallRector.php

```diff
@@ -0,0 +1,259 @@
+use PhpParser\Node\Expr\FuncCall;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\New_;
+use PhpParser\Node\Expr\PropertyFetch;
+use PhpParser\Node\Expr\StaticCall;
+use PhpParser\Node\Expr\StaticPropertyFetch;
+use PhpParser\Node\Expr\Variable;
+use PhpParser\Node\Name;
+use PhpParser\Node\Stmt\Return_;
+use PHPStan\Analyser\MutatingScope;
+use PHPStan\Analyser\Scope;
+use PHPStan\Broker\Broker;
+use PHPStan\Broker\FunctionNotFoundException;
+use PHPStan\Reflection\ParameterReflection;
+use PHPStan\Reflection\ParametersAcceptor;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use Rector\CodingStyle\Naming\ClassNaming;
+use Rector\NodeTypeResolver\Node\AttributeKey;
+use Rector\Php70\Dto\VariableAssignPair;
+use Rector\Rector\AbstractRector;
+use Rector\RectorDefinition\CodeSample;
+use Rector\RectorDefinition\RectorDefinition;
+
+/**
+ * @see https://www.php.net/manual/en/migration70.incompatible.php
+ *
+ * @see \Rector\Php70\Tests\Rector\FuncCall\NonVariableToVariableOnFunctionCallRector\NonVariableToVariableOnFunctionCallRectorTest
+ */
+final class NonVariableToVariableOnFunctionCallRector extends AbstractRector
+{
+    /**
+     * @var string
+     */
+    private const DEFAULT_VARIABLE_NAME = 'tmp';
+
+    /**
+     * @var Broker
+     */
+    private $broker;
+
+    /**
+     * @var ClassNaming
+     */
+    private $classNaming;
+
+    public function __construct(Broker $broker, ClassNaming $classNaming)
+    {
+        $this->broker = $broker;
+        $this->classNaming = $classNaming;
+    }
+
+    public function getDefinition(): RectorDefinition
+    {
+        return new RectorDefinition(
+            'Transform non variable like arguments to variable where a function or method expects an argument passed by reference',
+            [new CodeSample('reset(a());', '$a = a(); reset($a);')]
+        );
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [FuncCall::class, MethodCall::class, StaticCall::class];
+    }
+
+    /**
+     * @param FuncCall|MethodCall|StaticCall $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        $scope = $node->getAttribute(AttributeKey::SCOPE);
+        if (! $scope instanceof MutatingScope) {
+            return null;
+        }
+
+        $arguments = $this->getNonVariableArguments($node, $scope);
+
+        if ($arguments === []) {
+            return null;
+        }
+
+        foreach ($arguments as $key => $argument) {
+            $replacements = $this->processArgument($argument, $scope);
+
+            $current = $node->getAttribute(AttributeKey::CURRENT_STATEMENT);
+            $this->addNodeBeforeNode($replacements->assign(), $current instanceof Return_ ? $current : $node);
+            $node->args[$key]->value = $replacements->variable();
+
+            $scope = $scope->assignExpression($replacements->variable(), $scope->getType($replacements->variable()));
```

Comment:
I never used this before. What is it for? When should we use it?

Reply:
It returns a new Scope from the previous Scope + the assigned expression (for example, if you pass a variable to it, the new scope will be aware of this variable).

I did this so that Scope is aware of newly created variables, so that `createCountedValueName()` returns a truly unique variable name.

Reply:
I see, thanks!

Maybe it would be more consistent to add type here:

```php
/** @var MutatingScope $scope */
```

Reply:
`assignExpression()` always returns a MutatingScope. No need to add a phpDoc IMHO.

Reply:
I meant it to have `instanceof Scope` above and here make the annotation for variable on the right:

```php
/** @var MutatingScope $scope */
$whatever = **$scope**->assignExpression(...);
```